src/goto-programs/replace_calls.cpp

```diff
@@ -0,0 +1,158 @@
+/// \param replacement_map: mapping between function names; a mapping f -> g
+///   indicates that calls to f should be replaced by calls to g
+void replace_callst::operator()(
+  goto_modelt &goto_model,
+  const replacement_mapt &replacement_map) const
+{
+  check_replacement_map(replacement_map);
+
+  const namespacet ns(goto_model.symbol_table);
+  goto_functionst &goto_functions = goto_model.goto_functions;
+
+  for(auto &p : goto_functions.function_map)
+  {
+    goto_functionst::goto_functiont &goto_function = p.second;
+    goto_programt &goto_program = goto_function.body;
+
+    (*this)(goto_program, goto_functions, ns, replacement_map);
+  }
+}
+
+void replace_callst::operator()(
+  goto_programt &goto_program,
+  const goto_functionst &goto_functions,
+  const namespacet &ns,
+  const replacement_mapt &replacement_map) const
+{
+  Forall_goto_program_instructions(it, goto_program)
+  {
+    goto_programt::instructiont &ins = *it;
+
+    if(!ins.is_function_call())
+      continue;
+
+    code_function_callt &cfc = to_code_function_call(ins.code);
+    exprt &function = cfc.function();
+
+    PRECONDITION(function.id() == ID_symbol);
+
+    symbol_exprt &se = to_symbol_expr(function);
+    const irep_idt &id = se.get_identifier();
+
+    auto f_it1 = goto_functions.function_map.find(id);
+
+    INVARIANT(
+      f_it1 != goto_functions.function_map.end(),
+      "Called functions need to be present");
+
+    replacement_mapt::const_iterator r_it = replacement_map.find(id);
+
+    if(r_it == replacement_map.end())
+      continue;
+
+    const irep_idt &new_id = r_it->second;
+
+    auto f_it2 = goto_functions.function_map.find(new_id);
+
+    if(f_it2 == goto_functions.function_map.end())
+      throw "Replacement function " + id2string(new_id) +
+        " needs to be present";
+
+    const goto_functionst::goto_functiont &cur_goto_function = f_it1->second;
+    const goto_functionst::goto_functiont &new_goto_function = f_it2->second;
+
+    if(!base_type_eq(cur_goto_function.type, new_goto_function.type, ns))
+      throw "Functions " + id2string(id) + " and " + id2string(new_id) +
+        " are not type-compatible";
+
+    // check that returns have not been removed
+    goto_programt::const_targett next_it = std::next(it);
+    if(next_it != goto_program.instructions.end() && next_it->is_assign())
+    {
+      const code_assignt &ca = to_code_assign(next_it->code);
+      const exprt &rhs = ca.rhs();
+
+      if(rhs.id() == ID_symbol)
+      {
+        const symbol_exprt &se = to_symbol_expr(rhs);
+        if(has_suffix(id2string(se.get_identifier()), RETURN_VALUE_SUFFIX))
+          throw "Returns must not be removed before replacing calls";
+      }
+    }
+
+    // Finally modify the call
+    function.type() = new_goto_function.type;
+    se.set_identifier(new_id);
+  }
+}
+
+replace_callst::replacement_mapt replace_callst::parse_replacement_list(
+  const replacement_listt &replacement_list) const
```

Comment:
See, I think the checking whether they are defined, checking types, etc. should be done here.